src/main/java/space/obminyashka/items_exchange/service/impl/LocationServiceImpl.java

```diff
@@ -130,33 +130,31 @@ public String createParsedLocationsFile(List<RawLocation> creatingData) throws I
 
     @Override
     public String createParsedLocsFile(List<LocationRaw> creatingData) throws IOException {
-        Set<Area> areas = new HashSet<>();
-        Set<District> districts = new HashSet<>();
-        Set<City> cities = new HashSet<>();
+        Map<String, Area> areas = new HashMap<>();
+        Map<String, District> districts = new HashMap<>();
+        Map<String, City> cities = new HashMap<>();
 
         for (LocationRaw location : creatingData) {
             if (!location.getAreaUa().isBlank() && !location.getAreaEn().isBlank()) {
```

Comment:
you are checking whether _areaEn_ is not empty **in the controller**, can you then check everything in the same place: area, district, city?

Reply:
done